contrib/platform/src/com/sun/jna/platform/win32/COM/util/ProxyObject.java

```diff
@@ -0,0 +1,627 @@
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import com.sun.jna.Pointer;
+import com.sun.jna.WString;
+import com.sun.jna.platform.win32.Guid;
+import com.sun.jna.platform.win32.Guid.IID;
+import com.sun.jna.platform.win32.Guid.REFIID;
+import com.sun.jna.platform.win32.Kernel32;
+import com.sun.jna.platform.win32.Kernel32Util;
+import com.sun.jna.platform.win32.OaIdl;
+import com.sun.jna.platform.win32.OaIdl.DISPID;
+import com.sun.jna.platform.win32.OaIdl.DISPIDByReference;
+import com.sun.jna.platform.win32.OaIdl.EXCEPINFO;
+import com.sun.jna.platform.win32.OleAuto;
+import com.sun.jna.platform.win32.OleAuto.DISPPARAMS;
+import com.sun.jna.platform.win32.Variant;
+import com.sun.jna.platform.win32.Variant.VARIANT;
+import com.sun.jna.platform.win32.Variant.VariantArg;
+import com.sun.jna.platform.win32.WTypes;
+import com.sun.jna.platform.win32.WinDef;
+import com.sun.jna.platform.win32.WinDef.DWORDByReference;
+import com.sun.jna.platform.win32.WinDef.LCID;
+import com.sun.jna.platform.win32.WinDef.UINT;
+import com.sun.jna.platform.win32.WinNT;
+import com.sun.jna.platform.win32.WinNT.HRESULT;
+import com.sun.jna.platform.win32.COM.COMException;
+import com.sun.jna.platform.win32.COM.COMUtils;
+import com.sun.jna.platform.win32.COM.ConnectionPoint;
+import com.sun.jna.platform.win32.COM.ConnectionPointContainer;
+import com.sun.jna.platform.win32.COM.Dispatch;
+import com.sun.jna.platform.win32.COM.IDispatch;
+import com.sun.jna.platform.win32.COM.IDispatchCallback;
+import com.sun.jna.platform.win32.COM.util.annotation.ComInterface;
+import com.sun.jna.platform.win32.COM.util.annotation.ComMethod;
+import com.sun.jna.platform.win32.COM.util.annotation.ComProperty;
+import com.sun.jna.ptr.IntByReference;
+import com.sun.jna.ptr.PointerByReference;
+
+/**
+ * Ole32.INSTANCE.CoInitialize must be called on the current thread before using
+ * this object
+ */
+public class ProxyObject implements InvocationHandler, com.sun.jna.platform.win32.COM.util.IDispatch,
+		IRawDispatchHandle {
+
+	public ProxyObject(Class<?> theInterface, IDispatch rawDispatch, Factory factory) {
+		this.unknownId = -1;
+		this.rawDispatch = rawDispatch;
+		this.comThread = factory.getComThread();
+		this.theInterface = theInterface;
+		this.factory = factory;
+		//make sure dispatch object knows we have a reference to it
+		// (for debug it is usefult to be able to see how many refs are present
+		int n = this.rawDispatch.AddRef();
+		this.getUnknownId(); //pre cache it
+		factory.register(this);
+	}
+	
+	//cached value of the IUnknown interface pointer
+	// Rules of COM state that querying for the IUnknown interface must return an identical pointer value
+	long unknownId;
+	long getUnknownId() {
+		if (-1 == this.unknownId) {
+			try {
+	
+				final PointerByReference ppvObject = new PointerByReference();
+	
+//				HRESULT hr = this.comThread.execute(new Callable<HRESULT>() {

```

Comment:
Could you explain this? Were you trying to remove the requirement of on which thread to execute the function on? Generally I think just like COM requires you execute `CoInitialize` on a specific thread it's OK to require it for JNA. That said, does the commented code accomplish the same?

Mostly I think it's worth either a comment or it should be deleted.